File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
#include <vector>

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        vector<int> vi(256, -1);
        int max_len = 0, start = -1;
        for (int i = 0; i != s.length(); i++) {
            if (vi[s[i]] > start)
                start = vi[s[i]];
            vi[s[i]] = i;
            max_len = max(max_len, i - start);
        }
        return max_len;
    }
};

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
import java.util.Arrays;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        int[] curr = new int[256];
        Arrays.fill(curr, -1);
        int maxLen = 0;
        int start = -1;
        for (int i = 0; i < s.length(); ++i) {
            char c = s.charAt(i);
            if (curr[c] > start)
                start = curr[c];
            curr[c] = i;
            maxLen = Math.max(maxLen, i - start);
        }
        return maxLen;
    }
}